File: cpp/graphAlgorithms/spath/spt_bm.cpp
#include "stdinc.h"
#include "List.h"
#include "Graph_wd.h"

namespace grafalgo {

/** Compute a shortest path tree using Bellman-Moore scanning algorithm.
 *  @param g is a directed graph with edge lengths
 *  @param s is the source vertex for the shortest path tree computation;
 *  if s=0, paths are computed from an imaginary extra vertex with a
 *  zero length edge to every other vertex
 *  @param pEdge is an array of parent pointers; on return pEdge[u] is the
 *  number of the edge connecting u to its parent in the shortest path tree
 *  @param d is array of distances; on return d[u] is the shortest path
 *  distance from s to u
 *  @return true on success, false if a negative cost cycle was encountered
 */
bool spt_bm(Graph_wd& g, vertex s, edge pEdge[], edgeLength d[]) {
	for (vertex v = 1; v <= g.n(); v++) { pEdge[v] = 0; d[v] = INT_MAX; }

	List q(g.n()); vertex last;
	if (s != 0) {
		d[s] = 0; q.addLast(s); last = s;
	} else {
		for (vertex v = 1; v <= g.n(); v++) {
			d[v] = 0; q.addLast(v);
		}
		last = g.n();
	}

	int pass = 0; int cnt = 0;
	while (!q.empty()) {
		vertex v = q.first(); q.removeFirst();
		for (edge e = g.firstOut(v); e != 0; e = g.nextOut(v,e)) {
			vertex w = g.head(e);
			if (d[v] + g.length(e) < d[w]) {
				if (pEdge[w] == 0) cnt++;
				d[w] = d[v] + g.length(e); pEdge[w] = e;
				if (!q.member(w)) q.addLast(w);
			}
		}
		if (v == last && !q.empty()) { pass++; last = q.last(); }
		if (pass == g.n()) return false;
	}
	return cnt == g.n();
}

} // ends namespace

File: js/graphAlgorithms/spath/sptBM.mjs
/** @file sptBM.mjs
 *
 *  @author Jon Turner
 *  @date 2021
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import { assert, EnableAssert as ea } from '../../common/Assert.mjs';
import List from '../../dataStructures/basic/List.mjs';

/** Compute shortest path tree of a graph using Bellman-Moore algorithm.
 *  @param g is a digraph with edge lengths, some possibly negative
 *  @param s is a "source vertex" in g; if 0, distances are computed
 *  from a dummy source vertex with a zero-length edge to every real vertex
 *  @param trace controls output of information about the internal  
 *  state of the computation; larger values produce more information
 *  @return a tuple [link, dist, ts, stats] where link[u] is the edge from
 *  the parent of u to u in the spt rooted at s (or 0 if u unreachable);
 *  dist[u] is the shortest path distance from vertex s to vertex u
 *  (or infinity if u unreachable), ts is a trace string and stats is a
 *  statistics object; if g has a negative cycle [] is returned
 */
export default function sptBM(g, s, trace=0) {
	let link = new Int32Array(g.n+1); let ts = ''; let err = '';
	let dist = new Array(g.n+1).fill(Infinity);

	let q = new List(g.n);
	if (s != 0) {
		q.enq(s); dist[s] = 0;
	} else {
		for (let u = 1; u <= g.n; u++) {
			q.enq(u); dist[u] = 0;
		}
	}
	let pass = 0; let last = q.last(); let steps = 0;
	if (trace) {
		ts += g.toString(1) + 'initial queue: ' + q + '\n\n' +
			  'selected vertex, distance, edge to parent, queue, ' +
			  'pass count\n';
	}
	while (!q.empty()) {
		let u = q.deq();
		for (let e = g.firstOutof(u); e; e = g.nextOutof(u,e)) {
			let v = g.head(e); steps++;
			if (dist[v] > dist[u] + g.length(e)) {
				dist[v] = dist[u] + g.length(e); link[v] = e;
				if (!q.contains(v)) q.enq(v);
			}
		}
		if (u == last && !q.empty()) { pass++; last = q.last(); }
		if (trace) {
			ts += g.x2s(u) + ' ' +
				  (dist[u] != Infinity ? dist[u] : '-') + ' ' +
				  (link[u] != 0 ? g.e2s(link[u]) : '-') + ' ' +
				  q + ' ' + pass + '\n';
		}
		if (pass >= g.n) return [];
	}
	return [link, dist, ts, { 'passCount': pass, 'stepCount': steps }];
}
